Introduce a custom text field called SuperTextField

matthew-carroll opened the pull request:
This PR introduces a custom implementation for a text field that's called `SuperTextField`.

This work is initially intended to help Turtle. After it meets Turtle's needs then it will be maintained for general consumption.

This PR adds a demo screen to the `super_editor` demo app that shows a handful of use-cases for `SuperTextField`.

![Screen Shot 2021-04-30 at 9 46 04 PM](https://user-images.githubusercontent.com/7259036/116771100-a54c6780-a9fd-11eb-8b9b-35674bcb9775.png)

matthew-carroll commented:
@salihgueler please have a look at this PR after you're done with #150 

I'd like to get #150 and this PR merged by sometime early in the day May 6, my time. That way I can push forward with remaining v0.1.0 work.

For this PR, I recommend running the example app and checking out the new demo that I created for `SuperTextField`. After that, if you'd like, I can walk through it with you.

salihgueler commented on super_editor/lib/src/infrastructure/selectable_text.dart:
```diff
@@ -222,22 +224,42 @@ class SelectableTextState extends State<SelectableText> implements TextLayout {
+      // This condition was added because getOffsetForCaret() was throwing
+      // an exception when debugNeedsLayout is true. It's unclear what we're
+      // supposed to do at our level to ensure that condition doesn't happen
+      // so until we figure it out, we'll just return a zero Offset.
```
👍🏼

salihgueler commented on super_editor/lib/src/infrastructure/text_layout.dart:
```diff
@@ -0,0 +1,96 @@
+  List<TextBox> getBoxesForSelection(TextSelection selection);
+
+  /// Returns a bounding [Rect] for the character at the given [position].
+  Rect getCharacterBox(TextPosition position);
```
I would rename this as it gives an impression to return `TextBox` and not a `Rect`

```suggestion
  Rect getCharacterRect(TextPosition position);
```

matthew-carroll replied:
Good point. I'm gonna change the return type to `TextBox` to give it symmetry with `getBoxesForSelection()`.

salihgueler commented on super_editor/lib/src/infrastructure/text_layout.dart:
```diff
@@ -0,0 +1,96 @@
+  /// Returns the `TextPosition` in the first line within this
+  /// `TextLayout` that is closest to the given `x`-value, or
+  /// -1 if the text is not laid out yet.
+  TextPosition getPositionInFirstLineAtX(double x);
```
What is the "X" here? Is it the position in X Cartesian axis?

matthew-carroll replied:
Yeah, the doc comment above the method refers to "x-value". That's the horizontal position. Is there another name that you would prefer?

salihgueler commented on super_editor/lib/src/infrastructure/text_layout.dart:
```diff
@@ -0,0 +1,96 @@
+
+typedef TextExpansion = TextSelection Function(String text, TextPosition startingPosition, TextAffinity affinity);
+
+const paragraphExpansionFilter = _paragraphExpansionFilter;
```
I am seeing this usage for the first time 😅  Could you please explain why don't we expose the function and do this operation instead?

matthew-carroll replied:
I think I may have done this because I wasn't sure if the function could be listed as `const`, but I think it can, so I'll remove the variable and just leave the function public.

salihgueler commented on super_editor/lib/src/infrastructure/composable_text.dart:
```diff
@@ -0,0 +1,55 @@
+  ///
+  /// Throws an exception if [textPosition] is not the right type
+  /// for this [TextComposable].
+  TextSelection getWordSelectionAt(dynamic textPosition);
```
I am not so sure if we should throw an exception without a proper definition of allowed types here. Can we put something to class documentation? I know we already have possibilities but if we let the user decide the type, then I don't think that throwing non-allowed types should be their concern more than ours.

matthew-carroll replied:
That was done so that something like a table could still be considered a `TextComposable`. It's mentioned in the class docs:

```
/// The text positions accepted by a [TextComposable] are [dynamic]
/// rather than [TextPosition]s because a use-case might include
/// complex text composition, like tables, which might choose to
/// index positions based on cell IDs, or row and column indices.
```

salihgueler reviewed (commented):
Overall it looks good 👍🏼  Great work 👏🏼  I have some questions and I was wondering if we should add tests to SuperTextField and how it is created. WDYT?

matthew-carroll commented on super_editor/lib/src/infrastructure/selectable_text.dart:
```diff
@@ -3,77 +3,79 @@ import 'package:flutter/rendering.dart';
-}
+import 'text_layout.dart';
+
+// /// Queries and calculates text position and offset details
```
That was an oversight. I'll remove the code in the next commit.

matthew-carroll commented:
@salihgueler I rebased against `main` and fixed some related issues. I made some changes based on your PR comments. And I added tests for the `SuperTextField` keyboard handlers and actions.

salihgueler reviewed (approved):
Looks good now, we can merge this. It can be used in our app as well :)
